### DISCUSSION TITLE
High latency observed with Sinatra API intermittently

### DISCUSSION THREAD

**Discussion by imnani** [sinatra/sinatra#1863] [Q&A]
We are observing high latencies intermittently with our APIs which use the Sinatra framework and puma web server(single mode). We don't see any issues/delays with puma worker threads accepting incoming requests and processing the requests. The issue primarily lies with the delay involved in returning a response from the Sinatra controller. We use New Relic for our observability needs and it clearly indicates that the latency is caused by the controller.
Rack acts as an interface between the Sinatra framework and the puma server and I suspect rack could be contributing to this issue too. We are observing this intermittent issue only during peak load. Any guidance could be appreciated.

**Library Versions:**
Sinatra: 2.2.2
Puma: 5.5.2
Rack: 2.24

**Comment by dentarg** [comment]
What is your controller doing? 😄 I would concentrate on that 😉

**Reply by imnani** [reply]
Service logs indicate that controller functionality is processed in a few milliseconds and the latency is involved when we return the response object from the controller and the client receives the response.

**Reply by dentarg** [reply]
Slow clients?

This can't be answered without knowing the actual work your app is doing 

What is high latency? You need to quantify it

**Reply by dentarg** [reply]
Locks?

Have you tried increasing worker and thread count in Puma?

I think you need to reproduce the issue outside your application to get help from the community

**Reply by imnani** [reply]
> This can't be answered without knowing the actual work your app is doing
> We primarily call several internal services and validate against that data.

> What is high latency? You need to quantify it
Our API usually responds within 2-3 seconds with a normal load of 1TPS. When we run load tests at a rate of 3-5TPS intermittently we are observing a response time of more than 30 seconds. We don't see any issue with puma worker threads availability or the web server accepting requests at load. As per service logs, request processing didn't take much time but we observed a huge time difference between returning a response from the controller and the response received by the APIGEE proxy layer.

We use New Relic for monitoring needs and the traces clearly indicate that the latency is involved with process(Sinatra/V2::**Controller#POST (unknown)). As this is happening intermittently we are not able to figure out the root cause of this issue. I am suspecting the issue could be with Rack middleware too.

**Reply by imnani** [reply]
> Locks?
> Are you referring to puma thread locks?
> Have you tried increasing worker and thread count in Puma?
> I tried this approach and didn't really yield any results. Any suggested configuration for workers and threads in Puma?

**Reply by jkowens** [reply]
> We use New Relic for monitoring needs and the traces clearly indicate that the latency is involved with process(Sinatra/V2::**Controller#POST (unknown)). As this is happening intermittently we are not able to figure out the root cause of this issue. I am suspecting the issue could be with Rack middleware too.

Have you built out some type of framework on top of Sinatra? `Sinatra/V2::**Controller` looks like a custom routing configuration. I also noticed it was a `POST`. Could there be latency communicating with a database or another system?

**Reply by imnani** [reply]
We didn't build any framework on top of Sinatra framework. The above one is not a routing its the process name which is causing high latency from New Relic APM. We verified the processes which involve DB operations and we don't see any issues with them.

**Reply by imnani** [reply]
@dentarg @jkowens Any idea about what could be causing this issue at the controller level while returning the response

**Reply by jkowens** [reply]
Sorry @imnani with such little information, there isn't much help to provide. Could you post a screenshot of a New Relic trace of a slow request?

**Reply by imnani** [reply]
<img width="1471" alt="NewRelic Trace" src="https://user-images.githubusercontent.com/11781764/217366330-3635e4d3-e002-47ed-8f04-15ec62a57dfb.png">

If you observe the new relic trace it indicates that most time is consumed for the process **Application code in (Nested/Controller/Sinatra/V2::OrderController/POST (unknown))**. As I mentioned earlier based on service logs we figured out that at the controller level logic processing was pretty fast but when we returned the response from the controller it took a while for it to reach to APIGEE layer.

**Reply by dentarg** [reply]
APIGEE is a remote service, right? Isn't it just the time it takes to communicate with it?

Sinatra does not add 17 seconds of overhead

**Reply by jkowens** [reply]
Beside taking another look at the application code, maybe look at the Apigee and nginx logs.

**Reply by imnani** [reply]
New Relic instrumentation is done only for our application code and from New Relic APM it is pretty evident that 17 seconds overhead is added at the Sinatra controller level.

**Reply by dentarg** [reply]
I think you need to figure out this one yourselves. Sinatra is open-source after all, so you have that going for you.
